File: apps/windows/src/FlashTeX.Protocol/TransferV1.cs
// name: TransferV1.cs
// purpose: The Nearby/bridge capture-transfer protocol — documents, anchors,
//   proposals, capture lifecycle (docs/contracts/transfer-v1.md). Ported from
//   apps/mac/Sources/FlashTeXProtocol/TransferV1.swift. Additive to runtime v1: the
//   same envelope, snake_case keys, zero-based end-exclusive UTF-8 byte offsets, and
//   `error {code, message}` replies. `capture_submit` reuses `RuntimeV1.CaptureSubmit`.
// author: Claude Sonnet 5
// date: 2026-09-13

using System.Text.Json.Serialization;

namespace FlashTeX.Protocol.TransferV1;

/// <summary>Bridge protocol size/identifier limits (transfer-v1.md).</summary>
public static class TransferV1Protocol
{
    /// <summary>Each line, including its newline, is at most this many bytes in either direction.</summary>
    public const int MaxLineBytes = 12 * 1024 * 1024;
    public const int MaxRequestIdBytes = 128;
    public const int MaxDocumentBytes = 8 * 1024 * 1024;
    public const int MaxImageBytes = 8 * 1024 * 1024;
    public const int MaxInstructionBytes = 4096;

    /// <summary><c>capture_convert</c>'s <c>supported_features</c>: at most this many entries.</summary>
    public const int MaxSupportedFeatures = 64;
    public const int MaxSupportedFeatureBytes = 128;

    /// <summary><c>capture_proposal</c>/journaled proposal <c>latex</c> byte bounds.</summary>
    public const int MinLatexBytes = 1;
    public const int MaxLatexBytes = 65536;

    /// <summary><c>ambiguities</c>/<c>required_dependencies</c> list and entry bounds.</summary>
    public const int MaxProposalListEntries = 32;
    public const int MaxProposalEntryBytes = 2048;

    /// <summary>1-128 ASCII alphanumerics, <c>-</c> or <c>_</c>.</summary>
    public static bool IsValidIdentifier(string id)
    {
        if (id.Length == 0 || ByteOffsets.Utf8ByteCount(id) > MaxRequestIdBytes)
        {
            return false;
        }
        foreach (char c in id)
        {
            bool ok = (c >= '0' && c <= '9') || (c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z') || c == '-' || c == '_';
            if (!ok)
            {
                return false;
            }
        }
        return true;
    }

    /// <summary>Normalized relative path: no leading <c>/</c>, no <c>\</c>, <c>:</c>, NUL, empty, <c>.</c> or <c>..</c> segments.</summary>
    public static bool IsValidRelativePath(string path)
    {
        if (path.Length == 0 || path.StartsWith('/') || path.Contains('\\') || path.Contains(':') || path.Contains('\0'))
        {
            return false;
        }
        foreach (var segment in path.Split('/'))
        {
            if (segment.Length == 0 || segment == "." || segment == "..")
            {
                return false;
            }
        }
        return true;
    }
}

/// <summary>Request type to reply type, exactly as <c>crates/bridge/src/main.rs</c> dispatches.</summary>
public enum Request
{
    DocumentOpen,
    DocumentEdit,
    DestinationPin,
    CaptureSubmit,
    CaptureConvert,
    CapturePrepareInsert,
    CaptureApplied,
    CaptureStatus,
    CaptureReject,
}

/// <summary>Wire strings for <see cref="Request"/> and its reply type.</summary>
public static class RequestExtensions
{
    public static string WireName(this Request request) => request switch
    {
        Request.DocumentOpen => "document_open",
        Request.DocumentEdit => "document_edit",
        Request.DestinationPin => "destination_pin",
        Request.CaptureSubmit => "capture_submit",
        Request.CaptureConvert => "capture_convert",
        Request.CapturePrepareInsert => "capture_prepare_insert",
        Request.CaptureApplied => "capture_applied",
        Request.CaptureStatus => "capture_status",
        Request.CaptureReject => "capture_reject",
        _ => throw new ArgumentOutOfRangeException(nameof(request)),
    };

    public static string ReplyType(this Request request) => request switch
    {
        Request.DocumentOpen => "document_opened",
        Request.DocumentEdit => "document_updated",
        Request.DestinationPin => "destination_pinned",
        Request.CaptureSubmit => "capture_received",
        Request.CaptureConvert => "capture_proposal",
        Request.CapturePrepareInsert => "capture_edit",
        Request.CaptureApplied => "capture_application_received",
        Request.CaptureStatus => "capture_status",
        Request.CaptureReject => "capture_rejected",
        _ => throw new ArgumentOutOfRangeException(nameof(request)),
    };
}

/// <summary>
/// <c>error</c> reply payload. Codes are stable strings such as <c>provider_disabled</c>.
/// Deliberately a plain DTO (not an <see cref="Exception"/>): the Swift source's
/// <c>ErrorPayload</c> conforms to <c>Error</c> only so it can be thrown/caught in
/// Swift, and an <see cref="Exception"/> subtype here would leak base-class members
/// (StackTrace, Data, ...) into the wire shape via default JSON contract resolution.
/// </summary>
public sealed record ErrorPayload(
    [property: JsonPropertyName("code")] string Code,
    [property: JsonPropertyName("message")] string Message);

/// <summary>An empty payload, e.g. <c>document_opened</c>'s reply.</summary>
public sealed record Empty;

// MARK: documents and destinations

public sealed record DocumentOpen(
    [property: JsonPropertyName("project_id")] string ProjectId,
    [property: JsonPropertyName("path")] string Path,
    [property: JsonPropertyName("revision")] int Revision,
    [property: JsonPropertyName("text")] string Text);

public sealed record DocumentEdit(
    [property: JsonPropertyName("project_id")] string ProjectId,
    [property: JsonPropertyName("path")] string Path,
    [property: JsonPropertyName("base_revision")] int BaseRevision,
    [property: JsonPropertyName("revision")] int Revision,
    [property: JsonPropertyName("start_byte")] int StartByte,
    [property: JsonPropertyName("end_byte")] int EndByte,
    [property: JsonPropertyName("replacement")] string Replacement);

public sealed record DocumentUpdated([property: JsonPropertyName("revision")] int Revision);

public sealed record DestinationPin(
    [property: JsonPropertyName("destination_id")] string DestinationId,
    [property: JsonPropertyName("project_id")] string ProjectId,
    [property: JsonPropertyName("path")] string Path,
    [property: JsonPropertyName("revision")] int Revision,
    [property: JsonPropertyName("start_byte")] int StartByte,
    [property: JsonPropertyName("end_byte")] int EndByte);

/// <summary>Immutable binding of a destination to the exact source it was pinned in.</summary>
public sealed record AnchorBinding(
    [property: JsonPropertyName("project_id")] string ProjectId,
    [property: JsonPropertyName("path")] string Path,
    [property: JsonPropertyName("revision")] int Revision,
    [property: JsonPropertyName("start_byte")] int StartByte,
    [property: JsonPropertyName("end_byte")] int EndByte,
    [property: JsonPropertyName("source_sha256")] string SourceSha256);

public sealed record Anchor(
    [property: JsonPropertyName("destination_id")] string DestinationId,
    [property: JsonPropertyName("project_id")] string ProjectId,
    [property: JsonPropertyName("path")] string Path,
    [property: JsonPropertyName("pinned_revision")] int PinnedRevision,
    [property: JsonPropertyName("current_revision")] int CurrentRevision,
    [property: JsonPropertyName("start_byte")] int StartByte,
    [property: JsonPropertyName("end_byte")] int EndByte,
    [property: JsonPropertyName("valid")] bool IsValid,
    [property: JsonPropertyName("binding")] AnchorBinding Binding);

// MARK: capture lifecycle

public sealed record CaptureReceived(
    [property: JsonPropertyName("capture_id")] string CaptureId,
    [property: JsonPropertyName("durable")] bool Durable,
    [property: JsonPropertyName("has_proposal")] bool HasProposal,
    [property: JsonPropertyName("applied")] bool Applied);

/// <summary><c>capture_convert</c> request payload.</summary>
public sealed record CaptureConvert
{
    [JsonConstructor]
    public CaptureConvert(string captureId, IReadOnlyList<string> supportedFeatures)
    {
        if (supportedFeatures.Count > TransferV1Protocol.MaxSupportedFeatures)
        {
            throw new ArgumentException(
                $"supported_features carries {supportedFeatures.Count} entries (limit {TransferV1Protocol.MaxSupportedFeatures})");
        }
        foreach (var feature in supportedFeatures)
        {
            if (ByteOffsets.Utf8ByteCount(feature) > TransferV1Protocol.MaxSupportedFeatureBytes)
            {
                throw new ArgumentException($"supported feature '{feature}' exceeds {TransferV1Protocol.MaxSupportedFeatureBytes} bytes");
            }
        }
        CaptureId = captureId;
        SupportedFeatures = supportedFeatures;
    }

    [JsonPropertyName("capture_id")]
    public string CaptureId { get; }

    [JsonPropertyName("supported_features")]
    public IReadOnlyList<string> SupportedFeatures { get; }
}

public sealed record CaptureId([property: JsonPropertyName("capture_id")] string Value);

public sealed record CapturePrepareInsert(
    [property: JsonPropertyName("capture_id")] string CaptureId,
    [property: JsonPropertyName("expected_revision")] int ExpectedRevision,
    [property: JsonPropertyName("approved")] bool Approved);

/// <summary>
/// A prepared, persisted edit. Preparation never edits source; the Mac verifies
/// every field against its buffer before applying it once.
/// </summary>
public sealed record CaptureEdit(
    [property: JsonPropertyName("capture_id")] string CaptureId,
    [property: JsonPropertyName("edit_id")] string EditId,
    [property: JsonPropertyName("project_id")] string ProjectId,
    [property: JsonPropertyName("path")] string Path,
    [property: JsonPropertyName("expected_revision")] int ExpectedRevision,
    [property: JsonPropertyName("start_byte")] int StartByte,
    [property: JsonPropertyName("end_byte")] int EndByte,
    [property: JsonPropertyName("removed_text")] string RemovedText,
    [property: JsonPropertyName("replacement")] string Replacement,
    [property: JsonPropertyName("document_before_sha256")] string DocumentBeforeSha256);

public sealed record CaptureApplied(
    [property: JsonPropertyName("capture_id")] string CaptureId,
    [property: JsonPropertyName("edit_id")] string EditId,
    [property: JsonPropertyName("new_revision")] int NewRevision);

/// <summary>Journaled proposal as it appears inside <c>capture_status</c> (no capture_id).</summary>
public sealed record Proposal
{
    [JsonConstructor]
    public Proposal(string latex, IReadOnlyList<string> ambiguities, IReadOnlyList<string> requiredDependencies)
    {
        int latexBytes = ByteOffsets.Utf8ByteCount(latex);
        if (latexBytes < TransferV1Protocol.MinLatexBytes || latexBytes > TransferV1Protocol.MaxLatexBytes)
        {
            throw new ArgumentException($"latex must be {TransferV1Protocol.MinLatexBytes}...{TransferV1Protocol.MaxLatexBytes} UTF-8 bytes");
        }
        ValidateEntryList(ambiguities, nameof(ambiguities));
        ValidateEntryList(requiredDependencies, nameof(requiredDependencies));
        Latex = latex;
        Ambiguities = ambiguities;
        RequiredDependencies = requiredDependencies;
    }

    private static void ValidateEntryList(IReadOnlyList<string> entries, string fieldName)
    {
        if (entries.Count > TransferV1Protocol.MaxProposalListEntries)
        {
            throw new ArgumentException($"{fieldName} carries {entries.Count} entries (limit {TransferV1Protocol.MaxProposalListEntries})");
        }
        foreach (var entry in entries)
        {
            if (ByteOffsets.Utf8ByteCount(entry) > TransferV1Protocol.MaxProposalEntryBytes)
            {
                throw new ArgumentException($"{fieldName} entry exceeds {TransferV1Protocol.MaxProposalEntryBytes} bytes");
            }
        }
    }

    [JsonPropertyName("latex")]
    public string Latex { get; }

    [JsonPropertyName("ambiguities")]
    public IReadOnlyList<string> Ambiguities { get; }

    [JsonPropertyName("required_dependencies")]
    public IReadOnlyList<string> RequiredDependencies { get; }
}

public sealed record AppliedEdit(
    [property: JsonPropertyName("edit_id")] string EditId,
    [property: JsonPropertyName("new_revision")] int NewRevision);

public sealed record CaptureStatus(
    [property: JsonPropertyName("capture_id")] string CaptureId,
    [property: JsonPropertyName("proposal")] [property: JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)] Proposal? Proposal,
    [property: JsonPropertyName("prepared")] [property: JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)] CaptureEdit? Prepared,
    [property: JsonPropertyName("applied")] [property: JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)] AppliedEdit? Applied,
    [property: JsonPropertyName("rejected")] bool Rejected);

File: apps/mac/Sources/FlashTeXProtocol/TransferV1.swift
import Foundation

/// Models for the capture bridge contract (docs/contracts/transfer-v1.md,
/// consumed at bridge commit b5ca96b). Additive to runtime v1: the same
/// `{protocol_version, id, type, payload}` envelope, snake_case keys, zero-based
/// end-exclusive UTF-8 byte offsets, and `error {code, message}` replies.
/// `capture_submit` reuses `RuntimeV1.CaptureSubmit`.
public enum TransferV1 {
    /// Each line, including its newline, is at most 12 MiB in either direction.
    public static let maxLineBytes = 12 * 1024 * 1024
    public static let maxRequestIDBytes = 128
    public static let maxDocumentBytes = 8 * 1024 * 1024
    public static let maxImageBytes = 8 * 1024 * 1024
    public static let maxInstructionBytes = 4096

    /// Request type → reply type, exactly as `crates/bridge/src/main.rs` dispatches.
    public enum Request: String, CaseIterable {
        case documentOpen = "document_open"
        case documentEdit = "document_edit"
        case destinationPin = "destination_pin"
        case captureSubmit = "capture_submit"
        case captureConvert = "capture_convert"
        case capturePrepareInsert = "capture_prepare_insert"
        case captureApplied = "capture_applied"
        case captureStatus = "capture_status"
        case captureReject = "capture_reject"

        public var replyType: String {
            switch self {
            case .documentOpen: "document_opened"
            case .documentEdit: "document_updated"
            case .destinationPin: "destination_pinned"
            case .captureSubmit: "capture_received"
            case .captureConvert: "capture_proposal"
            case .capturePrepareInsert: "capture_edit"
            case .captureApplied: "capture_application_received"
            case .captureStatus: "capture_status"
            case .captureReject: "capture_rejected"
            }
        }
    }

    /// `error` reply payload. Codes are stable strings such as `provider_disabled`.
    public struct ErrorPayload: Codable, Equatable, Error {
        public var code: String
        public var message: String
        public init(code: String, message: String) { self.code = code; self.message = message }
    }

    public struct Empty: Codable, Equatable { public init() {} }

    // MARK: documents and destinations

    public struct DocumentOpen: Codable, Equatable {
        public var projectId: String
        public var path: String
        public var revision: Int
        public var text: String
        enum CodingKeys: String, CodingKey { case projectId = "project_id", path, revision, text }
        public init(projectId: String, path: String, revision: Int, text: String) {
            self.projectId = projectId; self.path = path; self.revision = revision; self.text = text
        }
    }

    public struct DocumentEdit: Codable, Equatable {
        public var projectId: String
        public var path: String
        public var baseRevision: Int
        public var revision: Int
        public var startByte: Int
        public var endByte: Int
        public var replacement: String
        enum CodingKeys: String, CodingKey {
            case projectId = "project_id", path, baseRevision = "base_revision", revision
            case startByte = "start_byte", endByte = "end_byte", replacement
        }
        public init(projectId: String, path: String, baseRevision: Int, revision: Int,
                    startByte: Int, endByte: Int, replacement: String) {
            self.projectId = projectId; self.path = path; self.baseRevision = baseRevision
            self.revision = revision; self.startByte = startByte; self.endByte = endByte
            self.replacement = replacement
        }
    }

    public struct DocumentUpdated: Codable, Equatable {
        public var revision: Int
        public init(revision: Int) { self.revision = revision }
    }

    /// How an anchor behaves under edits (additive `destination_pin.mode`).
    /// `fixed` is the explicit ⌘⌥P pin and the old contract: an edit at or
    /// across it invalidates it. `caret` is the Mac's automatic destination —
    /// the caret, wherever it is when the capture is inserted: it follows
    /// edits like a caret, is never invalidated by one, may be re-pinned under
    /// the same id anywhere, and treats the capture's `base_revision` as
    /// informational. See docs/contracts/transfer-v1.md.
    public enum AnchorMode: String, Codable, Equatable {
        case fixed, caret
    }

    public struct DestinationPin: Codable, Equatable {
        public var destinationId: String
        public var projectId: String
        public var path: String
        public var revision: Int
        public var startByte: Int
        public var endByte: Int
        /// Omitted (nil) means `fixed` to every bridge, old or new.
        public var mode: AnchorMode?
        enum CodingKeys: String, CodingKey {
            case destinationId = "destination_id", projectId = "project_id", path, revision
            case startByte = "start_byte", endByte = "end_byte", mode
        }
        public init(destinationId: String, projectId: String, path: String, revision: Int, startByte: Int, endByte: Int,
                    mode: AnchorMode? = nil) {
            self.destinationId = destinationId; self.projectId = projectId; self.path = path
            self.revision = revision; self.startByte = startByte; self.endByte = endByte; self.mode = mode
        }
    }

    /// Immutable binding of a destination to the exact source it was pinned in.
    public struct AnchorBinding: Codable, Equatable {
        public var projectId: String
        public var path: String
        public var revision: Int
        public var startByte: Int
        public var endByte: Int
        public var sourceSha256: String
        enum CodingKeys: String, CodingKey {
            case projectId = "project_id", path, revision, startByte = "start_byte", endByte = "end_byte"
            case sourceSha256 = "source_sha256"
        }
        public init(projectId: String, path: String, revision: Int, startByte: Int, endByte: Int, sourceSha256: String) {
            self.projectId = projectId; self.path = path; self.revision = revision
            self.startByte = startByte; self.endByte = endByte; self.sourceSha256 = sourceSha256
        }
    }

    public struct Anchor: Codable, Equatable {
        public var destinationId: String
        public var projectId: String
        public var path: String
        public var pinnedRevision: Int
        public var currentRevision: Int
        public var startByte: Int
        public var endByte: Int
        public var valid: Bool
        public var binding: AnchorBinding
        /// Absent from a bridge that predates `AnchorMode` (then `fixed`).
        public var mode: AnchorMode?
        enum CodingKeys: String, CodingKey {
            case destinationId = "destination_id", projectId = "project_id", path
            case pinnedRevision = "pinned_revision", currentRevision = "current_revision"
            case startByte = "start_byte", endByte = "end_byte", valid, binding, mode
        }
        public var isCaret: Bool { mode == .caret }
        public init(destinationId: String, projectId: String, path: String, pinnedRevision: Int, currentRevision: Int,
                    startByte: Int, endByte: Int, valid: Bool, binding: AnchorBinding, mode: AnchorMode? = nil) {
            self.destinationId = destinationId; self.projectId = projectId; self.path = path
            self.pinnedRevision = pinnedRevision; self.currentRevision = currentRevision
            self.startByte = startByte; self.endByte = endByte; self.valid = valid; self.binding = binding; self.mode = mode
        }
    }

    // MARK: capture lifecycle

    public struct CaptureReceived: Codable, Equatable {
        public var captureId: String
        public var durable: Bool
        public var hasProposal: Bool
        public var applied: Bool
        enum CodingKeys: String, CodingKey {
            case captureId = "capture_id", durable, hasProposal = "has_proposal", applied
        }
        public init(captureId: String, durable: Bool, hasProposal: Bool, applied: Bool) {
            self.captureId = captureId; self.durable = durable; self.hasProposal = hasProposal; self.applied = applied
        }
    }

    public struct CaptureConvert: Codable, Equatable {
        public var captureId: String
        public var supportedFeatures: [String]
        enum CodingKeys: String, CodingKey { case captureId = "capture_id", supportedFeatures = "supported_features" }
        public init(captureId: String, supportedFeatures: [String]) {
            self.captureId = captureId; self.supportedFeatures = supportedFeatures
        }
    }

    public struct CaptureID: Codable, Equatable {
        public var captureId: String
        enum CodingKeys: String, CodingKey { case captureId = "capture_id" }
        public init(captureId: String) { self.captureId = captureId }
    }

    /// What the Mac puts around the journaled proposal at approval (additive
    /// `capture_prepare_insert.wrap`): the delimiters and line breaks the
    /// caret's context calls for. The bridge journals it with the prepared
    /// edit; `replacement == prefix + proposal + suffix`. `kind` is a label
    /// (`display_math`, `inline_math`, `as_is`) the bridge does not interpret.
    public struct InsertionWrap: Codable, Equatable {
        public var prefix: String
        public var suffix: String
        public var kind: String?
        public init(prefix: String, suffix: String, kind: String? = nil) {
            self.prefix = prefix; self.suffix = suffix; self.kind = kind
        }
        public func applied(to proposal: String) -> String { prefix + proposal + suffix }
    }

    public struct CapturePrepareInsert: Codable, Equatable {
        public var captureId: String
        public var expectedRevision: Int
        public var approved: Bool
        /// Omitted (nil) is the old request; an older bridge ignores nothing
        /// because nothing is sent.
        public var wrap: InsertionWrap?
        enum CodingKeys: String, CodingKey {
            case captureId = "capture_id", expectedRevision = "expected_revision", approved, wrap
        }
        public init(captureId: String, expectedRevision: Int, approved: Bool, wrap: InsertionWrap? = nil) {
            self.captureId = captureId; self.expectedRevision = expectedRevision; self.approved = approved; self.wrap = wrap
        }
    }

    /// A prepared, persisted edit. Preparation never edits source; the Mac
    /// verifies every field against its buffer before applying it once.
    public struct CaptureEdit: Codable, Equatable {
        public var captureId: String
        public var editId: String
        public var projectId: String
        public var path: String
        public var expectedRevision: Int
        public var startByte: Int
        public var endByte: Int
        public var removedText: String
        /// `wrap.prefix + <journaled proposal> + wrap.suffix`.
        public var replacement: String
        public var documentBeforeSha256: String
        /// The wrap the edit was prepared with (absent when none was sent, or
        /// from a bridge / journal that predates it). Carried to the edit
        /// ledger as-is so its record and the bridge's agree.
        public var wrap: InsertionWrap?
        enum CodingKeys: String, CodingKey {
            case captureId = "capture_id", editId = "edit_id", projectId = "project_id", path
            case expectedRevision = "expected_revision", startByte = "start_byte", endByte = "end_byte"
            case removedText = "removed_text", replacement, documentBeforeSha256 = "document_before_sha256", wrap
        }
        public init(captureId: String, editId: String, projectId: String, path: String, expectedRevision: Int,
                    startByte: Int, endByte: Int, removedText: String, replacement: String, documentBeforeSha256: String,
                    wrap: InsertionWrap? = nil) {
            self.captureId = captureId; self.editId = editId; self.projectId = projectId; self.path = path
            self.expectedRevision = expectedRevision; self.startByte = startByte; self.endByte = endByte
            self.removedText = removedText; self.replacement = replacement; self.documentBeforeSha256 = documentBeforeSha256
            self.wrap = wrap
        }
    }

    public struct CaptureApplied: Codable, Equatable {
        public var captureId: String
        public var editId: String
        public var newRevision: Int
        enum CodingKeys: String, CodingKey { case captureId = "capture_id", editId = "edit_id", newRevision = "new_revision" }
        public init(captureId: String, editId: String, newRevision: Int) {
            self.captureId = captureId; self.editId = editId; self.newRevision = newRevision
        }
    }

    /// Journaled proposal as it appears inside `capture_status` (no capture_id).
    public struct Proposal: Codable, Equatable {
        public var latex: String
        public var ambiguities: [String]
        public var requiredDependencies: [String]
        enum CodingKeys: String, CodingKey { case latex, ambiguities, requiredDependencies = "required_dependencies" }
    }

    public struct AppliedEdit: Codable, Equatable {
        public var editId: String
        public var newRevision: Int
        enum CodingKeys: String, CodingKey { case editId = "edit_id", newRevision = "new_revision" }
    }

    public struct CaptureStatus: Codable, Equatable {
        public var captureId: String
        public var proposal: Proposal?
        public var prepared: CaptureEdit?
        public var applied: AppliedEdit?
        public var rejected: Bool
        enum CodingKeys: String, CodingKey { case captureId = "capture_id", proposal, prepared, applied, rejected }
    }

    // MARK: validation helpers (mirror the bridge's limits so bad requests fail locally)

    /// 1–128 ASCII alphanumerics, `-` or `_`.
    public static func isValidIdentifier(_ id: String) -> Bool {
        let u = id.utf8
        guard !u.isEmpty, u.count <= 128 else { return false }
        return u.allSatisfy { ($0 >= 0x30 && $0 <= 0x39) || ($0 >= 0x41 && $0 <= 0x5A) || ($0 >= 0x61 && $0 <= 0x7A) || $0 == 0x2D || $0 == 0x5F }
    }

    /// Normalized relative path: no leading `/`, no `\`, `:`, NUL, empty, `.` or `..` segments.
    public static func isValidRelativePath(_ path: String) -> Bool {
        guard !path.isEmpty, !path.hasPrefix("/"), !path.contains("\\"), !path.contains(":"), !path.contains("\0") else { return false }
        return !path.split(separator: "/", omittingEmptySubsequences: false).contains { $0.isEmpty || $0 == "." || $0 == ".." }
    }
}
